crates/shielded_token/src/vp.rs

```diff
@@ -978,4 +988,157 @@ mod shielded_token_tests {
+        tx.push_default_inner_tx();
+        let BatchedTx { tx, cmt } = tx.batch_first_tx();
+
+        // Write the conflicting keys
+        let amount = Amount::native_whole(100);
+        let _ = state
+            .write_log_mut()
+            .write(&balance_key, amount.serialize_to_vec())
+            .unwrap();
+        let token_map = TokenMap::new();
+        let _ = state
+            .write_log_mut()
+            .write(&token_map_key, token_map.serialize_to_vec())
+            .unwrap();
+
+        let gas_meter = RefCell::new(VpGasMeter::new_from_tx_meter(
+            &TxGasMeter::new(u64::MAX),
+        ));
+        let (vp_vp_cache, _vp_cache_dir) = vp_cache();
+        let ctx = Ctx::new(
+            &MASP,
+            &state,
+            &tx,
+            &cmt,
+            &tx_index,
+            &gas_meter,
+            &keys_changed,
+            &verifiers,
+            vp_vp_cache,
+        );
+
+        assert!(matches!(
+            MaspVp::validate_tx(
+                &ctx,
+                &tx.batch_ref_tx(&cmt),
+                &keys_changed,
+                &verifiers
+            ),
+            Err(Error::SimpleMessage(
+                "Cannot simultaneously do governance proposal and MASP \
+                 transfer"
+            ))
+        ));
+    }
+
+    proptest! {
+        // Changing no MASP keys at all is allowed
+        #[test]
+        fn test_no_masp_op_accepted(src_key in arb_key().prop_filter("MASP key", |key| !is_masp_key(key))) {
+            let mut state = TestState::default();
+            namada_parameters::init_test_storage(&mut state).unwrap();
+            let keys_changed = BTreeSet::from([src_key.clone()]);
+            let verifiers = Default::default();
+
+            let tx_index = TxIndex::default();
+            let mut tx = Tx::from_type(namada_tx::data::TxType::Raw);
+            tx.push_default_inner_tx();
+            let BatchedTx { tx, cmt } = tx.batch_first_tx();
+
+            // Write some random value
+            let _ = state
+                .write_log_mut()
+                .write(&src_key, "test".serialize_to_vec())
+                .unwrap();
+
+            let gas_meter = RefCell::new(VpGasMeter::new_from_tx_meter(
+                &TxGasMeter::new(u64::MAX),
+            ));
+            let (vp_vp_cache, _vp_cache_dir) = vp_cache();
+            let ctx = Ctx::new(
+                &MASP,
+                &state,
+                &tx,
+                &cmt,
+                &tx_index,
+                &gas_meter,
+                &keys_changed,
+                &verifiers,
+                vp_vp_cache,
+            );
+
+            assert!(MaspVp::validate_tx(
+                &ctx,
+                &tx.batch_ref_tx(&cmt),
+                &keys_changed,
+                &verifiers
+            )
+            .is_ok());
+        }
+
+        // Changing unknown masp keys is not allowed
+        #[test]
```

Comment:
fmt doesn't work well inside the proptest macro. This will have to be beautified manually